impeller/renderer/backend/metal/texture_mtl.mm

```diff
@@ -4,120 +4,295 @@
+      return false;
+    }
+
+    [encoder generateMipmapsForTexture:texture_];
+    mipmap_generated_ = true;
+
+    return true;
   }
 
-  is_wrapped_ = wrapped;
-  is_valid_ = true;
-}
+  // |Texture|
+  void SetLabel(std::string_view label) override {
+    [texture_ setLabel:@(label.data())];
+  }
+  // |Texture|
+  bool OnSetContents(const uint8_t* contents,
+                     size_t length,
+                     size_t slice) override {
+    if (!IsValid() || !contents || is_wrapped_) {
+      return false;
+    }
 
-std::shared_ptr<TextureMTL> TextureMTL::Wrapper(
-    TextureDescriptor desc,
-    id<MTLTexture> texture,
-    std::function<void()> deletion_proc) {
-  if (deletion_proc) {
-    return std::shared_ptr<TextureMTL>(
-        new TextureMTL(desc, texture, true),
-        [deletion_proc = std::move(deletion_proc)](TextureMTL* t) {
-          deletion_proc();
-          delete t;
-        });
+    const auto& desc = GetTextureDescriptor();
+
+    // Out of bounds access.
+    if (length != desc.GetByteSizeOfBaseMipLevel()) {
+      return false;
+    }
+
+    const auto region =
+        MTLRegionMake2D(0u, 0u, desc.size.width, desc.size.height);
+    [texture_ replaceRegion:region                            //
+                mipmapLevel:0u                                //
+                      slice:slice                             //
+                  withBytes:contents                          //
+                bytesPerRow:desc.GetBytesPerRow()             //
+              bytesPerImage:desc.GetByteSizeOfBaseMipLevel()  //
+    ];
+
+    return true;
   }
-  return std::shared_ptr<TextureMTL>(new TextureMTL(desc, texture, true));
-}
 
-TextureMTL::~TextureMTL() = default;
+  // |Texture|
+  bool OnSetContents(std::shared_ptr<const fml::Mapping> mapping,
+                     size_t slice) override {
+    // Metal has no threading restrictions. So we can pass this data along to
+    // the client rendering API immediately.
+    return OnSetContents(mapping->GetMapping(), mapping->GetSize(), slice);
+  }
 
-void TextureMTL::SetLabel(std::string_view label) {
-  [texture_ setLabel:@(label.data())];
-}
+  // |Texture|
+  bool IsValid() const override { return is_valid_; }
 
-// |Texture|
-bool TextureMTL::OnSetContents(std::shared_ptr<const fml::Mapping> mapping,
-                               size_t slice) {
-  // Metal has no threading restrictions. So we can pass this data along to the
-  // client rendering API immediately.
-  return OnSetContents(mapping->GetMapping(), mapping->GetSize(), slice);
-}
+  // |Texture|
+  ISize GetSize() const override {
+    return {static_cast<ISize::Type>(texture_.width),
+            static_cast<ISize::Type>(texture_.height)};
+  }
+
+  BetterNameTextureMTL(const BetterNameTextureMTL&) = delete;
+
+  BetterNameTextureMTL& operator=(const BetterNameTextureMTL&) = delete;
+};
+
+/// @brief A class that holds the mutable state of the DrawableTextureMTL.
+class DrawableHolder {
```

Comment:
I'd move these to their own TU. Generally prefer classes in their own TUs. Even for ones that are ostensibly "internal" to others for ease of code navigation IMO.

Reply:
Moved this to lazy_drawable_holder.h